#### Konwledge Sharing Content

<img src="https://pandas.pydata.org/static/img/pandas.svg" width="250">

# <center> Important Pandas Functions

#### By - [Bhanu Pratap Singh](https://www.linkedin.com/in/bpst/)

In [8]:
# importing packages
import pandas as pd

## Import CSV with `pd.read_csv()` function

In [13]:
# reading dataset - csv file (option 1)
df = pd.read_csv('datasets/world-happiness-report-2021.csv')
df

Country name  Regional indicator  Ladder score  \
0        Finland      Western Europe         7.842   
1        Denmark      Western Europe         7.620   
2    Switzerland      Western Europe         7.571   
3        Iceland      Western Europe         7.554   
4    Netherlands      Western Europe         7.464   
..           ...                 ...           ...   
144      Lesotho  Sub-Saharan Africa         3.512   
145     Botswana  Sub-Saharan Africa         3.467   
146       Rwanda  Sub-Saharan Africa         3.415   
147     Zimbabwe  Sub-Saharan Africa         3.145   
148  Afghanistan          South Asia         2.523   

     Standard error of ladder score  upperwhisker  lowerwhisker  \
0                             0.032         7.904         7.780   
1                             0.035         7.687         7.552   
2                             0.036         7.643         7.500   
3                             0.059         7.670         7.438   
4                             0.027         7.518         7.410   
..                              ...           ...           ...   
144                           0.120         3.748         3.276   
145                           0.074         3.611         3.322   
146                           0.068         3.548         3.282   
147                           0.058         3.259         3.030   
148                           0.038         2.596         2.449   

     Logged GDP per capita  Social support  Healthy life expectancy  \
0                   10.775           0.954                   72.000   
1                   10.933           0.954                   72.700   
2                   11.117           0.942                   74.400   
3                   10.878           0.983                   73.000   
4                   10.932           0.942                   72.400   
..                     ...             ...                      ...   
144                  7.926           0.787                   48.700   
145                  9.782           0.784                   59.269   
146                  7.676           0.552                   61.400   
147                  7.943           0.750                   56.201   
148                  7.695           0.463                   52.493   

     Freedom to make life choices  Generosity  Perceptions of corruption  \
0                           0.949      -0.098                      0.186   
1                           0.946       0.030                      0.179   
2                           0.919       0.025                      0.292   
3                           0.955       0.160                      0.673   
4                           0.913       0.175                      0.338   
..                            ...         ...                        ...   
144                         0.715      -0.131                      0.915   
145                         0.824      -0.246                      0.801   
146                         0.897       0.061                      0.167   
147                         0.677      -0.047                      0.821   
148                         0.382      -0.102                      0.924   

     Ladder score in Dystopia  Explained by: Log GDP per capita  \
0                        2.43                             1.446   
1                        2.43                             1.502   
2                        2.43                             1.566   
3                        2.43                             1.482   
4                        2.43                             1.501   
..                        ...                               ...   
144                      2.43                             0.451   
145                      2.43                             1.099   
146                      2.43                             0.364   
147                      2.43                             0.457   
148                      2.43                  

In [14]:
# reading dataset - csv file (option 2)
# Note - if we try to read this csv without encoding, we will get this error
# 'utf-8' codec can't decode byte 0xe9 in position 2776: invalid continuation byte
# for fixing such error we need to set encoding type
df = pd.read_csv('datasets/asian-restaurants.csv', encoding='latin-1')
df

case        restaurant  price  gReviews  gRating  ddReviews  ddRating  \
0       2          Sun Shui      2       324      4.5          0       0.0   
1       4    Chen Yang Li B      2       468      4.2          0       0.0   
2       6       Dragon Star      2       126      4.2          0       0.0   
3      11  Siam Orchid Thai      2       417      4.2          0       0.0   
4      12    Splendid Sushi      2       135      4.9          0       0.0   
..    ...               ...    ...       ...      ...        ...       ...   
98     85        Wok N Roll      2       107      4.3       2900       4.6   
99     61    Thousand Crane      2       176      3.5       3700       4.6   
100    50   Kisaki Japanese      2       590      4.4       3800       4.7   
101    41  Cheng Du Kitchen      2        31      4.5       5900       4.7   
102   103       Jade Garden      2        96      4.2          0       0.0   

     ddFee  yReviews  yRating   latitude  longitude        town  
0     0.00       154      4.0  42.919826 -71.461006     Bedford  
1     0.00        83      3.0  43.170056 -71.535558         Bow  
2     0.00        28      3.5  43.200392 -71.532658     Concord  
3     0.00        80      3.5  43.204658 -71.535469     Concord  
4     0.00         8      4.5  43.204005 -71.536732     Concord  
..     ...       ...      ...        ...        ...         ...  
98    0.99        27      4.0  42.753885 -71.462696      Nashua  
99    1.99        99      3.0  42.992559 -71.463695  Manchester  
100   1.99       112      4.0  42.987353 -71.462839  Manchester  
101   3.99        41      3.5  43.016959 -71.449888  Manchester  
102   0.00        26      4.0  43.048281 -71.434783    Hooksett  

[103 rows x 13 columns]

## Explore Data

### explore `shape` attribute - tells dimensionality of dataset

In [19]:
# check dataframe shape
df.shape

(103, 13)

this dataset has `103 rows` and `13 columns`

### explore `head()` function - to preview dataset

In [15]:
# head function - by default select top 5 rows
df.head()

case        restaurant  price  gReviews  gRating  ddReviews  ddRating  \
0     2          Sun Shui      2       324      4.5          0       0.0   
1     4    Chen Yang Li B      2       468      4.2          0       0.0   
2     6       Dragon Star      2       126      4.2          0       0.0   
3    11  Siam Orchid Thai      2       417      4.2          0       0.0   
4    12    Splendid Sushi      2       135      4.9          0       0.0   

   ddFee  yReviews  yRating   latitude  longitude     town  
0    0.0       154      4.0  42.919826 -71.461006  Bedford  
1    0.0        83      3.0  43.170056 -71.535558      Bow  
2    0.0        28      3.5  43.200392 -71.532658  Concord  
3    0.0        80      3.5  43.204658 -71.535469  Concord  
4    0.0         8      4.5  43.204005 -71.536732  Concord

In [16]:
# head function - select first 7 rows
df.head(7)

case        restaurant  price  gReviews  gRating  ddReviews  ddRating  \
0     2          Sun Shui      2       324      4.5          0       0.0   
1     4    Chen Yang Li B      2       468      4.2          0       0.0   
2     6       Dragon Star      2       126      4.2          0       0.0   
3    11  Siam Orchid Thai      2       417      4.2          0       0.0   
4    12    Splendid Sushi      2       135      4.9          0       0.0   
5    13   Szechuan Garden      2       386      4.1          0       0.0   
6    14        Tea Garden      2       342      4.2          0       0.0   

   ddFee  yReviews  yRating   latitude  longitude     town  
0    0.0       154      4.0  42.919826 -71.461006  Bedford  
1    0.0        83      3.0  43.170056 -71.535558      Bow  
2    0.0        28      3.5  43.200392 -71.532658  Concord  
3    0.0        80      3.5  43.204658 -71.535469  Concord  
4    0.0         8      4.5  43.204005 -71.536732  Concord  
5    0.0        39      3.5  43.259296 -71.579884  Concord  
6    0.0        49      3.0  43.209848 -71.537063  Concord

### explore `tail()` function

In [17]:
# tail function - by default select last 5 rows
df.tail()

case        restaurant  price  gReviews  gRating  ddReviews  ddRating  \
98     85        Wok N Roll      2       107      4.3       2900       4.6   
99     61    Thousand Crane      2       176      3.5       3700       4.6   
100    50   Kisaki Japanese      2       590      4.4       3800       4.7   
101    41  Cheng Du Kitchen      2        31      4.5       5900       4.7   
102   103       Jade Garden      2        96      4.2          0       0.0   

     ddFee  yReviews  yRating   latitude  longitude        town  
98    0.99        27      4.0  42.753885 -71.462696      Nashua  
99    1.99        99      3.0  42.992559 -71.463695  Manchester  
100   1.99       112      4.0  42.987353 -71.462839  Manchester  
101   3.99        41      3.5  43.016959 -71.449888  Manchester  
102   0.00        26      4.0  43.048281 -71.434783    Hooksett

In [18]:
# tail function - select last 2 rows
df.tail(2)

case        restaurant  price  gReviews  gRating  ddReviews  ddRating  \
101    41  Cheng Du Kitchen      2        31      4.5       5900       4.7   
102   103       Jade Garden      2        96      4.2          0       0.0   

     ddFee  yReviews  yRating   latitude  longitude        town  
101   3.99        41      3.5  43.016959 -71.449888  Manchester  
102   0.00        26      4.0  43.048281 -71.434783    Hooksett

## Data Types

In [20]:
# check dataframe data types
df.dtypes

case            int64
restaurant     object
price           int64
gReviews        int64
gRating       float64
ddReviews       int64
ddRating      float64
ddFee         float64
yReviews        int64
yRating       float64
latitude      float64
longitude     float64
town           object
dtype: object

In [21]:
# check specific column data type
df['restaurant'].dtype

dtype('O')

In [22]:
# check specific column data type
df.ddReviews.dtype

dtype('int64')

In [23]:
# check specific column data type
df['latitude'].dtype

dtype('float64')

### Sunsetting Data - `loc` and `iloc`

In [24]:
# select rows by index (note - index begins from 0)
df.loc[7:14]

case         restaurant  price  gReviews  gRating  ddReviews  ddRating  \
7     15       China Garden      1       144      3.9          0       0.0   
8     19         Sumo Sushi      2       327      4.4          0       0.0   
9     21        Golden Leaf      2        78      3.8          0       0.0   
10    22       Happy Family      2       291      4.3          0       0.0   
11    23              Osaka      2       192      4.0          0       0.0   
12    24       Asian Breeze      2       386      4.4          0       0.0   
13    25     Dynamite Sushi      2       249      4.4          0       0.0   
14    30  Bangkok Thai Food      2       192      4.7          0       0.0   

    ddFee  yReviews  yRating   latitude  longitude            town  
7     0.0        74      3.5  42.898300 -71.335473           Derry  
8     0.0        74      4.0  42.890518 -71.324262           Derry  
9     0.0        21      3.0  42.891259 -71.139307  East Hampstead  
10    0.0        23      3.0  43.230217 -71.361415           Epsom  
11    0.0        50      3.0  42.887691 -71.149027       Hampstead  
12    0.0        87      4.0  43.053215 -71.442233        Hooksett  
13    0.0       150      4.0  42.759067 -71.430858          Hudson  
14    0.0       111      4.5  42.866402 -71.349207     Londonderry

In [25]:
# select single cell value
df.loc[10, 'restaurant']

'Happy Family'

In [27]:
# select single cell by indexing row and column value
# row index - 10, column index - 1
df.iloc[10,1]

'Happy Family'

## explore `pd.to_csv()` function - save dataframe in a csv file format

In [28]:
# add test column to dataframe
df['test_column'] = 1

In [31]:
# view dataframe
df.head()

case        restaurant  price  gReviews  gRating  ddReviews  ddRating  \
0     2          Sun Shui      2       324      4.5          0       0.0   
1     4    Chen Yang Li B      2       468      4.2          0       0.0   
2     6       Dragon Star      2       126      4.2          0       0.0   
3    11  Siam Orchid Thai      2       417      4.2          0       0.0   
4    12    Splendid Sushi      2       135      4.9          0       0.0   

   ddFee  yReviews  yRating   latitude  longitude     town  test_column  
0    0.0       154      4.0  42.919826 -71.461006  Bedford            1  
1    0.0        83      3.0  43.170056 -71.535558      Bow            1  
2    0.0        28      3.5  43.200392 -71.532658  Concord            1  
3    0.0        80      3.5  43.204658 -71.535469  Concord            1  
4    0.0         8      4.5  43.204005 -71.536732  Concord            1

In [32]:
# save dataframe to csv
df.to_csv('datasets/asian-restaurants-output.csv', index=False)